src/core/pointcloud/qgspointcloudlayer.cpp

```diff
@@ -341,12 +361,30 @@ void QgsPointCloudLayer::setDataSource( const QString &dataSource, const QString
     }
   }
 
-  connect( mDataProvider.get(), &QgsPointCloudDataProvider::dataChanged, this, &QgsPointCloudLayer::dataChanged );
-
   emit dataSourceChanged();
   triggerRepaint();
 }
 
+void QgsPointCloudLayer::onPointCloudIndexLoaded()
+{
+  if ( mRenderer )
+  {
+    // Renderer is already set only in case
+    // we generated the point cloud index
+    // as a background job. In this case
```

Comment:
Having this logic based on mRenderer seems a bit fragile to me...

Loading of a point cloud index should be a fairly fast synchronous operation, where we don't need to emit signals... I would suggest to maybe instead have 1. enumeration with states: { not indexed, indexing, indexed }, and 2. signal to tell change of the state. This would have hopefully give better feedback also to other bits of the qgis environment - for example, layer tree panel could show indicators for the point cloud layers that are being indexed (or waiting in queue to be indexed). If we had a setting to disable auto-indexing of point cloud layers, then users could maybe even click these indicators to actually start indexing...

Now going a bit off-topic - it would be good to probably limit number of layers that get simultaneously indexed to max. 1 (at least initially) - indexing may take multiple threads and drag'n'drop of several point cloud files may make the computer unresponsive (and/or more likely to run out of memory).